File: android/app/src/main/java/sol/vortexx/android/rooms/impl/HttpRoomsRepository.kt
package sol.vortexx.android.rooms.impl

import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable
import sol.vortexx.android.crypto.api.Aead
import sol.vortexx.android.crypto.api.SecureRandomProvider
import sol.vortexx.android.crypto.util.Hex
import sol.vortexx.android.db.dao.RoomDao
import sol.vortexx.android.db.entities.RoomEntity
import sol.vortexx.android.net.impl.VortexHttpClient
import sol.vortexx.android.rooms.api.RefreshResult
import sol.vortexx.android.rooms.api.RoomResult
import sol.vortexx.android.rooms.api.RoomsRepository
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Rooms feature — thin wrapper over `/api/rooms/*`. Each method maps
 * 1-to-1 onto a server endpoint, with local DB upsert on success. The
 * heavy crypto (ECIES wrap of the new room key) stays here instead of
 * the view model — screens only deal with id+name.
 */
@Singleton
class HttpRoomsRepository @Inject constructor(
    private val http: VortexHttpClient,
    private val dao: RoomDao,
    private val aead: Aead,                       // referenced in Wave 11
    private val random: SecureRandomProvider,     // ditto
) : RoomsRepository {

    override fun observe(): Flow<List<RoomEntity>> = dao.observeAll()

    override suspend fun refresh(): RefreshResult = runCatching {
        val resp = http.client.get("api/rooms/my")
        if (!resp.status.isSuccess()) return RefreshResult.Error("http_${resp.status.value}")
        val body = resp.body<List<RoomDto>>()
        dao.upsertAll(body.map(RoomDto::toEntity))
        RefreshResult.Ok(body.size)
    }.getOrElse { RefreshResult.Error(it.message ?: "io_error") }

    override suspend fun create(name: String, isPrivate: Boolean): RoomResult {
        val req = CreateRoomReq(
            name       = name,
            is_private = isPrivate,
            // Key material — ECIES payload filled in by Wave 9's key-agreement step.
            // For Wave 8 we send an empty shell; the server rejects if missing.
            public_room_key_hex = if (!isPrivate) Hex.encode(random.nextBytes(32)) else null,
        )
        return runCatching {
            val resp = http.client.post("api/rooms") {
                contentType(ContentType.Application.Json); setBody(req)
            }
            if (!resp.status.isSuccess()) {
                RoomResult.Error("http_${resp.status.value}", resp.bodyAsText().take(200))
            } else {
                val dto = resp.body<RoomDto>()
                dao.upsert(dto.toEntity())
                RoomResult.Ok(dto.id)
            }
        }.getOrElse { RoomResult.Error("io", it.message ?: "request failed") }
    }

    override suspend fun joinByInvite(inviteCode: String): RoomResult = runCatching {
        val resp = http.client.post("api/rooms/join/$inviteCode")
        if (!resp.status.isSuccess()) return@runCatching RoomResult.Error(
            "http_${resp.status.value}", resp.bodyAsText().take(200),
        )
        val dto = resp.body<RoomDto>()
        dao.upsert(dto.toEntity())
        RoomResult.Ok(dto.id)
    }.getOrElse { RoomResult.Error("io", it.message ?: "request failed") }

    override suspend fun leave(roomId: Long): Boolean = runCatching {
        val resp = http.client.delete("api/rooms/$roomId/leave")
        if (resp.status.isSuccess()) { dao.delete(roomId); true } else false
    }.getOrDefault(false)

    // ── DTOs ───────────────────────────────────────────────────────────

    @Serializable
    private data class CreateRoomReq(
        val name: String,
        val is_private: Boolean,
        val public_room_key_hex: String? = null,
    )

    @Serializable
    private data class RoomDto(
        val id: Long,
        val name: String,
        val description: String = "",
        val invite_code: String,
        val is_private: Boolean = false,
        val is_channel: Boolean = false,
        val is_dm: Boolean = false,
        val avatar_emoji: String = "\uD83D\uDCAC",
        val member_count: Int = 0,
    ) {
        fun toEntity() = RoomEntity(
            id = id, name = name, description = description, inviteCode = invite_code,
            isPrivate = is_private, isChannel = is_channel, isDm = is_dm,
            avatarEmoji = avatar_emoji, memberCount = member_count,
        )
    }
}

File: ios/Modules/Sources/Rooms/impl/HttpRoomsRepository.swift
import Foundation
import Net
import DB
import VortexCrypto
import Identity

/// Creation flow (matches Kotlin `HttpRoomsRepository`):
///   1. Generate random 32-byte AES room key.
///   2. ECIES-encrypt to creator's own X25519 pubkey.
///   3. POST {encrypted_room_key, public_room_key_hex?} → /api/rooms.
///   4. Cache plaintext key locally so next send/receive skip a fetch.
public final class HttpRoomsRepository: RoomsRepository {

    private let http: HttpClient
    private let rooms: RoomDao
    private let roomKeys: RoomKeyDao
    private let aead: Aead
    private let keyAgreement: KeyAgreement
    private let kdf: Kdf
    private let random: SecureRandomProvider
    private let identity: IdentityRepository

    public init(
        http: HttpClient,
        rooms: RoomDao,
        roomKeys: RoomKeyDao,
        aead: Aead,
        keyAgreement: KeyAgreement,
        kdf: Kdf,
        random: SecureRandomProvider,
        identity: IdentityRepository,
    ) {
        self.http = http
        self.rooms = rooms
        self.roomKeys = roomKeys
        self.aead = aead
        self.keyAgreement = keyAgreement
        self.kdf = kdf
        self.random = random
        self.identity = identity
    }

    public func observe() -> AsyncStream<[RoomRecord]> { rooms.observeAll() }

    public func refresh() async -> RefreshResult {
        do {
            let list = try await http.send(.get("api/rooms/my"), [RoomDto].self)
            try await rooms.upsertAll(list.map(\.toRecord))
            return .ok(count: list.count)
        } catch HttpError.status(let code, _) {
            return .error("http_\(code)")
        } catch {
            return .error((error as NSError).localizedDescription)
        }
    }

    public func create(name: String, isPrivate: Bool) async -> RoomResult {
        do {
            let me = try await identity.createOrLoad()
            let roomKey = random.nextBytes(32)

            // ECIES(room_key, my_pubkey). Wire format matches the server's
            // EncryptedRoomKey.ciphertext — nonce(12) || ct || tag(16).
            let eph = keyAgreement.generateKeyPair()
            let shared = try keyAgreement.agree(myPrivate: eph.privateKey, theirPublic: me.x25519.publicKey)
            let aesKey = try kdf.derive(ikm: shared, info: Data("vortex/ecies".utf8), length: 32)
            let ctPacked = try aead.encrypt(key: aesKey, plaintext: roomKey)

            let req = CreateReq(
                name: name,
                is_private: isPrivate,
                encrypted_room_key: EciesDto(
                    ephemeral_pub: Hex.encode(eph.publicKey),
                    ciphertext: Hex.encode(ctPacked),
                ),
                public_room_key_hex: isPrivate ? nil : Hex.encode(roomKey),
            )
            let dto = try await http.send(try HttpRequest.postJson("api/rooms", body: req), RoomDto.self)
            try await rooms.upsert(dto.toRecord)
            try await roomKeys.upsert(RoomKeyRecord(
                roomId: dto.id,
                keyHex: Hex.encode(roomKey),
                algorithm: "aes-256-gcm",
                source: isPrivate ? "ecies" : "public",
                createdAt: Int64(Date().timeIntervalSince1970 * 1000),
            ))
            return .ok(roomId: dto.id)
        } catch HttpError.status(let code, let body) {
            return .error(code: "http_\(code)", message: body)
        } catch {
            return .error(code: "io", message: (error as NSError).localizedDescription)
        }
    }

    public func joinByInvite(_ code: String) async -> RoomResult {
        do {
            let dto = try await http.send(
                HttpRequest(method: .POST, path: "api/rooms/join/\(code)"),
                RoomDto.self,
            )
            try await rooms.upsert(dto.toRecord)
            return .ok(roomId: dto.id)
        } catch HttpError.status(let code, let body) {
            return .error(code: "http_\(code)", message: body)
        } catch {
            return .error(code: "io", message: (error as NSError).localizedDescription)
        }
    }

    public func leave(_ roomId: Int64) async -> Bool {
        do {
            _ = try await http.send(.delete("api/rooms/\(roomId)/leave"))
            try await rooms.delete(roomId)
            return true
        } catch { return false }
    }

    // ── DTOs ───────────────────────────────────────────────────────────

    private struct EciesDto: Codable {
        let ephemeral_pub: String
        let ciphertext: String
    }
    private struct CreateReq: Encodable {
        let name: String
        let is_private: Bool
        let encrypted_room_key: EciesDto
        let public_room_key_hex: String?
    }
    private struct RoomDto: Decodable {
        let id: Int64
        let name: String
        let description: String?
        let invite_code: String
        let is_private: Bool?
        let is_channel: Bool?
        let is_dm: Bool?
        let avatar_emoji: String?
        let member_count: Int?

        var toRecord: RoomRecord {
            RoomRecord(
                id: id, name: name, desc: description ?? "",
                inviteCode: invite_code,
                isPrivate: is_private ?? false,
                isChannel: is_channel ?? false,
                isDm: is_dm ?? false,
                avatarEmoji: avatar_emoji ?? "💬",
                memberCount: member_count ?? 0,
            )
        }
    }
}
